README.md
# NSL Competitive Mod
A Natural Selection 2 balance and feature mod used by the [ENSL](https://www.ensl.org).

# Mission statement
The vision of the Natural Selection League (NSL) Competitive Modification (CompMod) is to enable fun, fair and balanced play in a 6 vs 6 environment, while remaining as accessible to new players as possible.
To do this the team commits to remaining transparent in all changes, to be open to discussion, feedback and criticism, and above all else, to strive to attain enjoyable play for all members of the competitive community, regardless of skill level. 

# Changes
For a full list of changes, read the documentation [here](https://enslcompmod.github.io/CompMod/index "CompMod Documentation").


docs/changes.md
# Changes between CompMod 3.5.6 and Vanilla 336
<br/>

# Alien
![alt text](https://wiki.naturalselection2.com/images/9/9d/Movement_Banner.png "Alien")

## BoneWall
* BoneWall is no longer flammable

## Drifters
* Reduced hover height to 1 from 1.2 so they don't float in Marine's faces
* Drifters cloaked by Shade hive upgrade will now uncloak from further away (to 5 from 1.5)
* ### Abilities
  * Enzyme
    * Reduced cooldown from 12 seconds to 1 second
  * Hallucinations
    * No longer affected by mucous
    * Can no longer have babblers attached
    * Reduced Onos HP to 100 eHP
  * Mucous
    * Reduced cooldown to 1 second from 12 seconds
    * Reduced cloud radius from 8 to 5 (reduces area by 60%)
    * Shield matches babbler shield values (except Skulk - remains at 15 HP)

## Eggs
* Hives no longer heal eggs, only embryos
* Lifeform egg drops removed

## Embryos
* Embryo HP scales depending on lifeform

## Fade
* ### Advanced Swipe
  * Replaces Stab
  * Upgrades the normal swipe damage to 81 from 75
  * Can be researched at Biomass 7
  * Costs 25 tres and takes 60 seconds to research
* ### Movement
  * Removed auto-crouch
  * Reduced ground slide duration to 1 second from 2 seconds
  * Blink no longer ignores the Fade's momentum
  * Added a miniumum speed of 15 to Blink. When changing direction you are guaranteed this speed as a minimum
  * Softcaps added to Fade speed:
    * 19 without Celerity
    * 20.5 with Celerity
  * Movement Breakdown:
    * Without Celerity
      * Initial Blink will give 15 speed from 16.25 speed,
      * Each subsequent blink will now give 2.5 speed
    * With Celerity
      * Initial Blink will give 15.6 speed from 18.28 speed,
      * Each subsequent blink will give 2.9 speed

## Gorge
* ### Babblers
  * Are flammable (die in one welder tick)
  * Reduced spawn rate to 3sec/babbler from 2.5sec/babbler
* ### BileBomb
  * Research changed to Biomass 2 from Biomass 3
* ### Web
  * HP gain per charge to 5 from 10
  * Web charges lowered to 0 from 3
  * Webbed duration lowered to 2.5 seconds from 5 seconds

## Lerk
* ### Movement
  * Increase strafe force to 8.3 from 7
  * Air friction value remains equal to vanilla friction (0.08) for 0.64 seconds after flapping
  * After 0.75 seconds, friction is scaled linearly from 0.08 to 0.5 over 3.25 seconds (total of 4 seconds)
  * This change minimizes the effectiveness of silent Lerk ambushes. A lerk will have to flap in order to quickly catch a marine
  * Lerk movement mid-flight while flapping often is unaffected
* ### Spikes
  * Damage increased to 6 from 5 (to 12 from 10 to players)
  * Spread increased to 3.8 radians from 3.6 radians
  * Size reduced to 45mm from 60mm
* ### Spores
  * Research changed to Biomass 6 from Biomass 5
  * Opacity of cloud lowered by 40%
* ### Umbra
  * Reduced opacity by 25%
* ### Vampirism Shield
  * Reduced leeched health amount to 6% from 8%

## Onos
* ### BoneShield
  * Add small initial cost to BoneShield
  * Moving with BoneShield active will consume BoneFuel faster
* ### Charge
  * Marine knockback removed

## Skulk
* ### BiteCone
  * Reduced width of cone from 1.2 to 0.8
* ### Leap
  * Energy cost increased to 55 from 45

## Structures
* ### Cyst
  * Build time increased to 6 seconds from 3.33 seconds
  * Shade hive cysts are now visible from further away (to 10 from 6)
* ### Shift
  * Echo
    * Echo cost for upgrades increased to 2 from 1
    * Echo cost for eggs reduced to 1 from 2
* ### Tunnels
  * Tunnels are infested tunnels by default
  * Tunnel cost changed to 8 tres from 6 tres
  * Decreased height check for tunnel placement

## Upgrades
* ### Adrenaline
  * Removed additional energy pool from Adrenaline
* ### Aura
  * Max range decreased to 24 from 30 (to 8 per veil from 10 per veil)
  * Only shows HP value on parasited players
* ### Camouflage
  * No longer fully cloaked while moving
* ### Carapace
  * Upgrade removed
* ### Neurotoxin
  * Replaces Focus
  * All Alien primary attacks will inflict a poison toxin, hurting Marines over time
  * Damage will tick once every second
  * Duration will be 1 second per Veil
  * Damage Values:
    * Skulk: 6
    * Gorge: 5
    * Lerk: 4
    * Fade: 8
    * Onos: 6
* ### Regeneration
  * Removed heal effect (visual and audio)
  * Reduced from 8% per tick to 6% per tick
* ### Vampirism
  * No longer triggers from friendly-fire damage
  * Works against exosuits
  * Now applies a shader to players that have Vampirism shield

# Marine
![alt text](https://wiki.naturalselection2.com/images/3/30/Marine_banner.png "Marine")

## ARCs
* Health lowered to 2200/400 from 2600/400

## Advanced Weapons
* Added new research 'Demolitions'
  * Researched on Advanced Armory
  * Research cost 15 tres
  * Research time 30 seconds
  * Unlocks Flamethrower and Grenade Launcher to purchase from the Advanced Armory

## MACs
* Cost reduced to 4 tres from 5 tres

## Medpacks
* Marines now keep the HoT effect even when they're full HP for the full duration of the Medpack. They cannot overheal.
* The result is that Marines can take damage after receiving a Medpack and still benefit from the HoT buff, even if they were already healed to full HP.

## Nanoshield
* Duration on players reduced to 2 seconds from 3 seconds

## Structures
* ### AdvancedArmory
  * Health decreased to 2000/200 from 3000/200
  * Research cost decreased to 15 tres from 25 tres
  * Research time decreased to 45 seconds from 90 seconds
* ### Observatory
  * Changed build time to 10 seconds from 15 seconds
* ### Prototype Lab
  * Cost reduced to 25 from 35
* ### Robotics Factory
  * Removed Armory requirement for ARC Factory upgrade
  * ARC Factory upgrade cost increased to 15 tres from 5 tres
* ### Sentry
  * Cost increased to 6 tres from 5 tres
  * Confusion from Spores
    * Duration increased to 8 seconds from 4 seconds
    * Time until next attack increased to 4 seconds from 2 seconds
* ### Sentry Battery

lua/ARC.lua
-- ======= Copyright (c) 2003-2012, Unknown Worlds Entertainment, Inc. All rights reserved. =======
--
-- lua\ARC.lua
--
--    Created by:   Charlie Cleveland (charlie@unknownworlds.com) and
--                  Max McGuire (max@unknownworlds.com)
--
-- AI controllable "tank" that the Commander can move around, deploy and use for long-distance
-- siege attacks.
--
-- ========= For more information, visit us at http://www.unknownworlds.com =====================

Script.Load("lua/ScriptActor.lua")
Script.Load("lua/Mixins/ClientModelMixin.lua")
Script.Load("lua/DoorMixin.lua")
Script.Load("lua/RagdollMixin.lua")
Script.Load("lua/LiveMixin.lua")
Script.Load("lua/UpgradableMixin.lua")
Script.Load("lua/PointGiverMixin.lua")
Script.Load("lua/AchievementGiverMixin.lua")
Script.Load("lua/TeamMixin.lua")
Script.Load("lua/GameEffectsMixin.lua")
Script.Load("lua/FlinchMixin.lua")
Script.Load("lua/OrdersMixin.lua")
Script.Load("lua/SelectableMixin.lua")
Script.Load("lua/MobileTargetMixin.lua")
Script.Load("lua/LOSMixin.lua")
Script.Load("lua/PathingMixin.lua")
Script.Load("lua/RepositioningMixin.lua")
Script.Load("lua/NanoShieldMixin.lua")
Script.Load("lua/SleeperMixin.lua")
Script.Load("lua/WeldableMixin.lua")
Script.Load("lua/TargetCacheMixin.lua")
Script.Load("lua/DissolveMixin.lua")
Script.Load("lua/DamageMixin.lua")
Script.Load("lua/CorrodeMixin.lua")
Script.Load("lua/MapBlipMixin.lua")
Script.Load("lua/UnitStatusMixin.lua")
Script.Load("lua/CommanderGlowMixin.lua")
Script.Load("lua/SupplyUserMixin.lua")
Script.Load("lua/CombatMixin.lua")
Script.Load("lua/IdleMixin.lua")
Script.Load("lua/WebableMixin.lua")
Script.Load("lua/ParasiteMixin.lua")
Script.Load("lua/RolloutMixin.lua")

class 'ARC' (ScriptActor)

ARC.kMapName = "arc"

ARC.kModelName = PrecacheAsset("models/marine/arc/arc.model")
local kAnimationGraph = PrecacheAsset("models/marine/arc/arc.animation_graph")

-- Animations
local kArcPitchParam = "arc_pitch"
local kArcYawParam = "arc_yaw"

ARC.kArcForwardTrackYawParam = "move_yaw"
ARC.kArcForwardTrackPitchParam = "move_pitch"
-- Balance
ARC.kHealth                 = kARCHealth
ARC.kStartDistance          = 4
ARC.kAttackDamage           = kARCDamage
ARC.kFireRange              = kARCRange         -- From NS1
ARC.kMinFireRange           = kARCMinRange
ARC.kSplashRadius           = 7
ARC.kUpgradedSplashRadius   = 13
ARC.kMoveSpeed              = 2.0
ARC.kCombatMoveSpeed        = 0.8
ARC.kFov                    = 360
ARC.kBarrelMoveRate         = 100
ARC.kMaxPitch               = 45
ARC.kMaxYaw                 = 180
ARC.kCapsuleHeight = .05
ARC.kCapsuleRadius = .5

ARC.kMode = enum( {'Stationary', 'Moving', 'Targeting', 'Destroyed'} )

ARC.kDeployMode = enum( { 'Undeploying', 'Undeployed', 'Deploying', 'Deployed' } )

ARC.kTurnSpeed = math.pi / 2 -- an ARC turns slowly
ARC.kMaxSpeedLimitAngle = math.pi / 36 -- 5 degrees
ARC.kNoSpeedLimitAngle = math.pi / 4 -- 24 degrees

if Server then
    Script.Load("lua/ARC_Server.lua")
end

local networkVars =
{
    -- ARCs can only fire when deployed and can only move when not deployed
    mode = "enum ARC.kMode",
    deployMode = "enum ARC.kDeployMode",
    
    barrelYawDegrees = "compensated float",
    barrelPitchDegrees = "compensated float",
    
    -- pose parameters for forward track (should be compensated??)
    forwardTrackYawDegrees = "float",
    forwardTrackPitchDegrees = "float",
    
    -- So we can update angles and pose parameters smoothly on client
    targetDirection = "vector",
}

AddMixinNetworkVars(BaseModelMixin, networkVars)
AddMixinNetworkVars(ClientModelMixin, networkVars)
AddMixinNetworkVars(LiveMixin, networkVars)
AddMixinNetworkVars(UpgradableMixin, networkVars)
AddMixinNetworkVars(GameEffectsMixin, networkVars)
AddMixinNetworkVars(FlinchMixin, networkVars)
AddMixinNetworkVars(TeamMixin, networkVars)
AddMixinNetworkVars(OrdersMixin, networkVars)
AddMixinNetworkVars(NanoShieldMixin, networkVars)
AddMixinNetworkVars(DissolveMixin, networkVars)
AddMixinNetworkVars(CorrodeMixin, networkVars)
AddMixinNetworkVars(LOSMixin, networkVars)
AddMixinNetworkVars(SelectableMixin, networkVars)
AddMixinNetworkVars(CombatMixin, networkVars)
AddMixinNetworkVars(IdleMixin, networkVars)
AddMixinNetworkVars(WebableMixin, networkVars)
AddMixinNetworkVars(ParasiteMixin, networkVars)

function ARC:OnCreate()

    ScriptActor.OnCreate(self)
    
    InitMixin(self, BaseModelMixin)
    InitMixin(self, ClientModelMixin)
    InitMixin(self, DoorMixin)
    InitMixin(self, LiveMixin)
    InitMixin(self, RagdollMixin)
    InitMixin(self, UpgradableMixin)
    InitMixin(self, GameEffectsMixin)
    InitMixin(self, FlinchMixin, { kPlayFlinchAnimations = true })
    InitMixin(self, TeamMixin)
    InitMixin(self, PointGiverMixin)
    InitMixin(self, AchievementGiverMixin)
    InitMixin(self, OrdersMixin, { kMoveOrderCompleteDistance = kAIMoveOrderCompleteDistance })
    InitMixin(self, PathingMixin)
    InitMixin(self, SelectableMixin)
    InitMixin(self, DissolveMixin)
    InitMixin(self, DamageMixin)
    InitMixin(self, CorrodeMixin)
    InitMixin(self, EntityChangeMixin)
    InitMixin(self, LOSMixin)
    InitMixin(self, CombatMixin)
    InitMixin(self, WebableMixin)
    InitMixin(self, ParasiteMixin)
    InitMixin(self, RolloutMixin)
    
    if Server then
    
        InitMixin(self, RepositioningMixin)
        InitMixin(self, SleeperMixin)
        
        self.targetPosition = nil
        self.targetedEntity = Entity.invalidId
        
    elseif Client then
        InitMixin(self, CommanderGlowMixin)
    end
    
    self.deployMode = ARC.kDeployMode.Undeployed
    
    self:SetLagCompensated(true)
    
end

function ARC:OnInitialized()

    ScriptActor.OnInitialized(self)
    
    InitMixin(self, WeldableMixin)
    InitMixin(self, NanoShieldMixin)
    
    self:SetModel(ARC.kModelName, kAnimationGraph)
    
    if Server then
    
        local angles = self:GetAngles()
        self.desiredPitch = angles.pitch
        self.desiredRoll = angles.roll
    
        InitMixin(self, MobileTargetMixin)
        InitMixin(self, SupplyUserMixin)
        
        -- TargetSelectors require the TargetCacheMixin for cleanup.
        InitMixin(self, TargetCacheMixin)
        
        -- Prioritize targetting non-Eggs first.
        self.targetSelector = TargetSelector():Init(
                self,
                ARC.kFireRange,
                false, 
                { kMarineStaticTargets, kMarineMobileTargets },
                { self.FilterTarget(self) },
                { function(target) return target:isa("Hive") end })

        
        self:SetPhysicsType(PhysicsType.Kinematic)
        
        -- Cannons start out mobile
        self:SetMode(ARC.kMode.Stationary)
        
        self.undeployedArmor = kARCArmor
        self.deployedArmor = kARCDeployedArmor
        
        -- This Mixin must be inited inside this OnInitialized() function.
        if not HasMixin(self, "MapBlip") then
            InitMixin(self, MapBlipMixin)
        end
    
        self.desiredForwardTrackPitchDegrees = 0
        
        InitMixin(self, InfestationTrackerMixin)
    
    elseif Client then
    
        self.lastModeClient = self.mode
        InitMixin(self, UnitStatusMixin)
        InitMixin(self, HiveVisionMixin)
    
    end
    
    self:SetUpdates(true, kRealTimeUpdateRate)
    
    InitMixin(self, IdleMixin)
    
end

function ARC:GetHealthbarOffset()
    return 0.7
end 

function ARC:GetPlayIdleSound()
    return self.deployMode == ARC.kDeployMode.Undeployed
end

function ARC:GetReceivesStructuralDamage()
    return true
end

function ARC:GetTurnSpeedOverride()
    return ARC.kTurnSpeed
end

function ARC:GetSpeedLimitAnglesOverride()
    return { ARC.kMaxSpeedLimitAngle, ARC.kNoSpeedLimitAngle }
end

function ARC:GetCanSleep()
    return self.mode == ARC.kMode.Stationary
end

function ARC:GetDeathIconIndex()
    return kDeathMessageIcon.ARC
end

--
-- Put the eye up 1 m.
--
function ARC:GetViewOffset()
    return self:GetCoords().yAxis * 1.0
end

function ARC:GetEyePos()
    return self:GetOrigin() + self:GetViewOffset()
end

function ARC:Deploy(commander)

    local queuedDeploy = commander ~= nil and commander.shiftDown

    if queuedDeploy then
    
        local lastOrder = self:GetLastOrder()        
        local orderOrigin = lastOrder ~=  nil and lastOrder:GetLocation() or self:GetOrigin()
        
        self:GiveOrder(kTechId.ARCDeploy, self:GetId(), orderOrigin, nil, false, false)
        
    else

        self:ClearOrders()
        self.deployMode = ARC.kDeployMode.Deploying
        self:TriggerEffects("arc_deploying")
    
    end

end

function ARC:UnDeploy()

end

function ARC:PerformActivation(techId, position, normal, commander)

    if techId == kTechId.ARCDeploy then
    
        self:Deploy(commander)
        return true, true
        
    elseif techId == kTechId.ARCUndeploy then
        
        if self:GetTarget() ~= nil then
            self:CompletedCurrentOrder()
        end
        
        self:SetMode(ARC.kMode.Stationary)
        
        self.deployMode = ARC.kDeployMode.Undeploying
        
        self:TriggerEffects("arc_stop_charge")
        self:TriggerEffects("arc_undeploying")
        
        return true, true
        
    end  
    
    self.targetPosition = nil
    
    return false, true
    
end

function ARC:GetTechAllowed(techId, techNode, player)

    local allowed, canAfford = ScriptActor.GetTechAllowed(self, techId, techNode, player)
    
    if self.deployMode == ARC.kDeployMode.Deployed and techId == kTechId.ARCUndeploy then
        allowed = true
    end
    
    return allowed, canAfford

end

function ARC:GetActivationTechAllowed(techId)

    if techId == kTechId.ARCDeploy then
        return self.deployMode == ARC.kDeployMode.Undeployed
    elseif techId == kTechId.Move then
        return self.deployMode == ARC.kDeployMode.Undeployed
    elseif techId == kTechId.ARCUndeploy then
        return self.deployMode == ARC.kDeployMode.Deployed
    elseif techId == kTechId.Stop then
        return self.mode == ARC.kMode.Moving or self.mode == ARC.kMode.Targeting
    end
    
    return true
    
end

function ARC:GetTechButtons(techId)

    return  { kTechId.Move, kTechId.Stop, kTechId.Attack, kTechId.None,
              kTechId.ARCDeploy, kTechId.ARCUndeploy, kTechId.None, kTechId.None }
              
end

function ARC:GetInAttackMode()
    return self.deployMode == ARC.kDeployMode.Deployed
end

function ARC:GetCanGiveDamageOverride()
    return true
end

function ARC:GetFov()
    return ARC.kFov
end

function ARC:OnOverrideDoorInteraction(inEntity)
    return true, 4
end

function ARC:GetEffectParams(tableParams)
    tableParams[kEffectFilterDeployed] = self:GetInAttackMode()
end

function ARC:FilterTarget()

    local attacker = self
    return function (target, targetPosition) return attacker:GetCanFireAtTargetActual(target, targetPosition) end
    
end

-- for marquee selection
function ARC:GetIsMoveable()
    return true
end

--
-- Do a complete check if the target can be fired on.
--
function ARC:GetCanFireAtTarget(target, targetPoint)    

    if target == nil then        
        return false
    end
    
    if not HasMixin(target, "Live") or not target:GetIsAlive() then
        return false
    end
    
    if not GetAreEnemies(self, target) then        
        return false
    end
    
    if not target.GetReceivesStructuralDamage or not target:GetReceivesStructuralDamage() then        
        return false
    end
    
    -- don't target eggs (they take only splash damage)
    if target:isa("Egg") or target:isa("Cyst") then
        return false
    end
    
    return self:GetCanFireAtTargetActual(target, targetPoint)
    
end

function ARC:GetCanBeUsed(player, useSuccessTable)
    useSuccessTable.useSuccess = false    
end

--
-- the checks made in GetCanFireAtTarget has already been made by the TargetCache, this
-- is the extra, actual target filtering done.
--
function ARC:GetCanFireAtTargetActual(target, targetPoint)    

    if not target.GetReceivesStructuralDamage or not target:GetReceivesStructuralDamage() then        
        return false
    end
    
    -- don't target eggs (they take only splash damage)
    if target:isa("Egg") or target:isa("Cyst") then
        return false
    end
    
    if not target:GetIsSighted() and not GetIsTargetDetected(target) then
        return false
    end
    
    local distToTarget = (target:GetOrigin() - self:GetOrigin()):GetLengthXZ()
    if (distToTarget > ARC.kFireRange) or (distToTarget < ARC.kMinFireRange) then
        return false
    end
    
    return true
    
end

function ARC:UpdateAngles(deltaTime)

    if not self:GetInAttackMode() or not self:GetIsAlive() then
        return
    end
    
    if self.mode == ARC.kMode.Targeting then
    
        if self.targetDirection then
        
            local yawDiffRadians = GetAnglesDifference(GetYawFromVector(self.targetDirection), self:GetAngles().yaw)
            local yawDegrees = DegreesTo360(math.deg(yawDiffRadians))    
            self.desiredYawDegrees = Clamp(yawDegrees, -ARC.kMaxYaw, ARC.kMaxYaw)
            
            local pitchDiffRadians = GetAnglesDifference(GetPitchFromVector(self.targetDirection), self:GetAngles().pitch)
            local pitchDegrees = DegreesTo360(math.deg(pitchDiffRadians))
            self.desiredPitchDegrees = -Clamp(pitchDegrees, -ARC.kMaxPitch, ARC.kMaxPitch)       
            
            self.barrelYawDegrees = Slerp(self.barrelYawDegrees, self.desiredYawDegrees, ARC.kBarrelMoveRate * deltaTime)
            
        end
        
    elseif self.deployMode == ARC.kDeployMode.Deployed or self.mode == ARC.kMode.Targeting then
    
        self.desiredYawDegrees = 0
        self.desiredPitchDegrees = 0
        
        self.barrelYawDegrees = Slerp(self.barrelYawDegrees, self.desiredYawDegrees, ARC.kBarrelMoveRate * deltaTime)
        
    end
    
    self.barrelPitchDegrees = Slerp(self.barrelPitchDegrees, self.desiredPitchDegrees, ARC.kBarrelMoveRate * deltaTime)
    
end

function ARC:OnUpdatePoseParameters()

    PROFILE("ARC:OnUpdatePoseParameters")
    
    self:SetPoseParam(kArcPitchParam, self.barrelPitchDegrees)
    self:SetPoseParam(kArcYawParam , self.barrelYawDegrees)
    self:SetPoseParam(ARC.kArcForwardTrackYawParam , self.forwardTrackYawDegrees)
    self:SetPoseParam(ARC.kArcForwardTrackPitchParam , self.forwardTrackPitchDegrees)
    
end

function ARC:OnUpdate(deltaTime)

    PROFILE("ARC:OnUpdate")
    
    ScriptActor.OnUpdate(self, deltaTime)
    
    if Server then
    
        self:UpdateOrders(deltaTime)
        self:UpdateSmoothAngles(deltaTime)

    end
    
    if self.mode ~= ARC.kMode.Stationary and self.mode ~= ARC.kMode.Moving and self.deployMode ~= ARC.kDeployMode.Deploying and self.mode ~= ARC.kMode.Destroyed then
        self:UpdateAngles(deltaTime)
    end
    
    if Client then
    
        if self.lastModeClient ~= self.mode then
            self:OnModeChangedClient(self.lastModeClient, self.mode)
        end
    
        self.lastModeClient = self.mode
    
    end
    
end

function ARC:OnModeChangedClient(oldMode, newMode)

    if oldMode == ARC.kMode.Targeting and newMode ~= ARC.kMode.Targeting then
        self:TriggerEffects("arc_stop_effects")
    end

end

function ARC:OnKill(attacker, doer, point, direction)

    self:TriggerEffects("arc_stop_effects")
    
    if Server then
    
        self:ClearTargetDirection()
        self:ClearOrders()
        
        self:SetMode(ARC.kMode.Destroyed)
        
    end 
  
end

function ARC:OnUpdateAnimationInput(modelMixin)

    PROFILE("ARC:OnUpdateAnimationInput")
    
    local activity = "none"
    if self.mode == ARC.kMode.Targeting and self.deployMode == ARC.kDeployMode.Deployed then
        activity = "primary"
    end
    modelMixin:SetAnimationInput("activity", activity)
    
    local deployed = self.deployMode == ARC.kDeployMode.Deploying or self.deployMode == ARC.kDeployMode.Deployed
    modelMixin:SetAnimationInput("deployed", deployed)
    
    local move = "idle"
    if self.mode == ARC.kMode.Moving and self.deployMode == ARC.kDeployMode.Undeployed then
        move = "run"
    end
    modelMixin:SetAnimationInput("move", move)
    
end

function ARC:GetShowHitIndicator()
    return false
end

function ARC:OnOverrideOrder(order)
    if order:GetType() == kTechId.Default then
        if self.deployMode == ARC.kDeployMode.Deployed then
            order:SetType(kTechId.Attack)
        else
            order:SetType(kTechId.Move)
        end
    end
end

function ARC:OnOrderGiven(order)
    if order ~= nil and (order:GetType() == kTechId.Attack or order:GetType() == kTechId.SetTarget) then
        local target = Shared.GetEntity(order:GetParam())
        if target then
            local dist = (self:GetOrigin() - target:GetOrigin()):GetLength()
            local valid = true
            if not HasMixin(target, "Live") or not target:GetIsAlive() then
                valid = false
            end
            if not GetAreEnemies(self, target) then        
                valid = false
            end
            if not target.GetReceivesStructuralDamage or not target:GetReceivesStructuralDamage() then        
                valid = false
            end
            if dist and valid and dist >= ARC.kMinFireRange and dist <= ARC.kFireRange then
                self.targetedEntity = order:GetParam()
                self.orderedEntity = order:GetParam()
            end
        end
    end
end

Shared.LinkClassToMap("ARC", ARC.kMapName, networkVars, true)
